kernel/libs/xarray/src/cursor.rs

```diff
@@ -0,0 +1,419 @@
+    /// potentially with creating new nodes.
+    fn continue_traverse_to_target_mut(&mut self) {
+        while !self.state.is_at_leaf() {
+            let (current_node, operation_offset) =
+                core::mem::take(&mut self.state).into_node().unwrap();
+
+            if current_node
+                .entry_with(self.guard, operation_offset)
+                .is_none()
+            {
+                let new_node = XNode::new(current_node.height().go_leaf(), operation_offset);
+
+                new_node.set_parent(&current_node);
+
+                let new_entry = Either::Left(Arc::new(new_node));
+                current_node.set_entry(self.guard, operation_offset, Some(new_entry));
+            }
+
+            let next_node = current_node
+                .deref_target()
+                .entry_with(self.guard, operation_offset)
+                .unwrap()
+                .left()
+                .unwrap();
+
+            self.0.state.move_to(next_node, self.0.index);
+        }
+    }
+
+    /// Traverses from the root node to the leaf node according to the target index,
+    /// potentially with creating new nodes.
+    fn expand_and_traverse_to_target(&mut self) {
+        if self.state.is_at_node() {
+            return;
+        }
+
+        let head = {
+            self.reserve(self.index);
+            self.xa.head.read_with(self.guard).unwrap()
+        };
+
+        self.0.state.move_to(head.left().unwrap(), self.0.index);
+        self.continue_traverse_to_target_mut();
+    }
+
+    /**** Public ****/
+
+    /// Stores a new `item` at the target index.
+    pub fn store(&mut self, item: P) {
+        self.expand_and_traverse_to_target();
+        let (node, operation_offset) = self.state.as_node().unwrap();
+        node.set_entry(self.guard, operation_offset, Some(Either::Right(item)));
+    }
+
+    /// Removes the item at the target index, and returns the removed item
+    /// if it previously exists.
+    pub fn remove(&mut self) -> Option<P::Ref<'a>> {
+        self.traverse_to_target();
+        self.state.as_node().and_then(|(node, off)| {
+            let res = node
+                .deref_target()
+                .entry_with(self.guard, off)
+                .and_then(|entry| entry.right());
+            node.set_entry(self.guard, off, None);
+            res
+        })
+    }
+
+    /// Sets the input `mark` for the item at the target index if the
+    /// target item exists, otherwise returns an error.
+    pub fn set_mark(&mut self, mark: M) -> ostd::Result<()> {
+        self.traverse_to_target();
+        self.state
+            .as_node()
+            .filter(|(node, off)| {
+                node.entry_with(self.guard, *off)
+                    .is_some_and(|entry| entry.is_right())
+            })
+            .map(|(node, off)| {
+                let mark_index = mark.into().index();
+                node.set_mark(self.guard, off, mark_index);
+            })
+            .ok_or(Error::AccessDenied)
```

Comment:
For basic data structures, we should define error types ourselves. See examples in the Rust standard library:  [`TryReserveError`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.try_reserve), [`FromVecWithNulError`](https://doc.rust-lang.org/std/ffi/struct.CString.html#method.from_vec_with_nul).

```rust
/// An error indicating that the mark cannot be set because the item does not exist.
struct SetMarkError;

    pub fn set_mark(&mut self, mark: M) -> Result<(), SetMarkError>;
    pub fn unset_mark(&mut self, mark: M) -> Result<(), SetMarkError>;
```